### ISSUE TITLE
auto-layout constraint warnings

### ISSUE THREAD

**Issue by armanshan12** [jessesquires/JSQMessagesViewController#1241]
Please help! If you scroll up and down, you would notice a huge lag and eventually the app crashes and I get the following in the console.

Library version(s) :octocat: 7.2.0
iOS version(s) :iphone: iOS 9
Devices/Simulators affected :iphone: iPhone 6
Complete steps to reproduce the issue :warning: scroll up and down and eventually it crashes
Link to a project that exhibits the issue, if possible fork the repo and modify the provided demo project :construction: Firebase example also crashed when I scrolled up and down on it, without changing any of their code.
Full crash log, if applicable :boom:

````
Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'layout attributes for supplementary item at index path (<NSIndexPath: 0xc000000000000016> {length = 2, path = 0 - 0}) changed from <JSQMessagesCollectionViewLayoutAttributes: 0x1310fa590> index path: (<NSIndexPath: 0xc000000000000016> {length = 2, path = 0 - 0}); element kind: (UICollectionElementKindSectionHeader); frame = (0 0; 375 32); zIndex = 10;  to <JSQMessagesCollectionViewLayoutAttributes: 0x1311073e0> index path: (<NSIndexPath: 0xc000000000000016> {length = 2, path = 0 - 0}); element kind: (UICollectionElementKindSectionHeader); frame = (0 0; 375 32); zIndex = -1;  without invalidating the layout'
````

**Comment by harlanhaskins** [comment]
Can you post some of your code?

**Comment by armanshan12** [comment]
I actually tried the Firebase example and that also crashed with the same error.

Here is my code:

    override func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int
    {
        return self.messages.count
    }

    override func collectionView(collectionView: JSQMessagesCollectionView!, messageDataForItemAtIndexPath indexPath: NSIndexPath!) -> JSQMessageData!
    {
        return self.messages[indexPath.row].chatMessage
    }
    
    override func collectionView(collectionView: JSQMessagesCollectionView!, messageBubbleImageDataForItemAtIndexPath indexPath: NSIndexPath!) -> JSQMessageBubbleImageDataSource!
    {
        let message = self.messages[indexPath.row]
        if message.author == PFUser.currentUser()! {
            return self.outgoingBubbleImageView
        }
        else {
            return self.incomingBubbleImageView
        }
    }
    
    override func collectionView(collectionView: JSQMessagesCollectionView!, avatarImageDataForItemAtIndexPath indexPath: NSIndexPath!) -> JSQMessageAvatarImageDataSource!
    {
        let gender = self.messages[indexPath.row].author["gender"] as! String
        let avatarName = gender == "female" ? "female-avatar-blue" : "male-avatar-blue"
        let avatar = JSQMessagesAvatarImage(avatarImage: nil, highlightedImage: nil, placeholderImage: UIImage(named: avatarName))
        
        let message = self.messages[indexPath.row]
        let author = message.author
        let userID = author["facebookID"] as! String
        let url = NSURL(string: "https://graph.facebook.com/\(userID)/picture")
        
        SDWebImageManager.sharedManager().downloadImageWithURL(url, options: .LowPriority, progress: nil) { (image, error, _, _, _) -> Void in
            avatar.avatarImage = image
            avatar.avatarHighlightedImage = image
            self.collectionView!.reloadItemsAtIndexPaths([indexPath])
        }
    
        return avatar
    }
    
    override func collectionView(collectionView: JSQMessagesCollectionView!, layout collectionViewLayout: JSQMessagesCollectionViewFlowLayout!, heightForCellTopLabelAtIndexPath indexPath: NSIndexPath!) -> CGFloat
    {
        if indexPath.item % 3 == 0 {
            return kJSQMessagesCollectionViewCellLabelHeightDefault
        }
        return 0.0
    }
    
    override func collectionView(collectionView: JSQMessagesCollectionView!, attributedTextForCellTopLabelAtIndexPath indexPath: NSIndexPath!) -> NSAttributedString!
    {
        if indexPath.item % 3 == 0 {
            let message = self.messages[indexPath.row]
            let date = message.createdAt ?? NSDate()
            return JSQMessagesTimestampFormatter.sharedFormatter().attributedTimestampForDate(date)
        }
        return nil
    }
    
    override func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell
    {
        let cell = super.collectionView(collectionView, cellForItemAtIndexPath: indexPath) as! JSQMessagesCollectionViewCell
        let message = self.messages[indexPath.row]
        
        
        cell.avatarImageView!.layer.cornerRadius = cell.avatarImageView!.frame.size.height / 2
        cell.avatarImageView!.layer.masksToBounds = true
        cell.avatarImageView!.layer.borderWidth = 0
        
        cell.messageBubbleTopLabel!.text = message.author!["firstname"] as? String
        cell.cellBottomLabel!.text = self.messages[indexPath.row].address
        
        if message.author == PFUser.currentUser()! {
            cell.textView!.textColor = UIColor.whiteColor()
        }
        else {
            cell.textView!.textColor = UIColor.blackColor()
        }
        
        return cell
    }
    
    override func collectionView(collectionView: JSQMessagesCollectionView!, layout collectionViewLayout: JSQMessagesCollectionViewFlowLayout!, heightForCellBottomLabelAtIndexPath indexPath: NSIndexPath!) -> CGFloat
    {
        let message = self.messages[indexPath.row]
        
        if message.author.objectId == PFUser.currentUser()!.objectId {
            if indexPath.row != self.messages.count-1 { //not last message
                let nextMessage = self.messages[indexPath.row+1]
                if message.author.objectId == nextMessage.author.objectId {
                    if let address = message.address, let nextAddress = nextMessage.address {
                        if address == nextAddress {
                            return 0
                        }
                    }
                }
            }
        }
        
        return self.heightForCellBottomLabelAtIndexPath(indexPath)
    }
    
    func heightForCellBottomLabelAtIndexPath(indexPath: NSIndexPath) -> CGFloat
    {
        if self.messages[indexPath.row].address != nil {
            return kJSQMessagesCollectionViewCellLabelHeightDefault
        }
        return 0.0
    }
    
    override func collectionView(collectionView: JSQMessagesCollectionView!, layout collectionViewLayout: JSQMessagesCollectionViewFlowLayout!, heightForMessageBubbleTopLabelAtIndexPath indexPath: NSIndexPath!) -> CGFloat
    {
        let message = self.messages[indexPath.row]
        
        if message.author.objectId! == PFUser.currentUser()!.objectId! {
            return 0.0
        }
        
        if indexPath.row == 0 { //first message
            return kJSQMessagesCollectionViewCellLabelHeightDefault
        }
        
        let prevMessage = self.messages[indexPath.row-1]
        if message.author.objectId == prevMessage.author.objectId {
            return 0
        }
        
        return kJSQMessagesCollectionViewCellLabelHeightDefault
    }

**Comment by harlanhaskins** [comment]
Your

```rust
override func collectionView(collectionView: JSQMessagesCollectionView!, layout collectionViewLayout: JSQMessagesCollectionViewFlowLayout!, heightForCellBottomLabelAtIndexPath indexPath: NSIndexPath!) -> CGFloat
```

calls

```rust
return self.heightForCellBottomLabelAtIndexPath(indexPath)
```

Which will infinitely recurse and crash.

**Comment by jessesquires** [comment]
Thanks @armanshan12 :smile: please review the [contribution guidelines](https://github.com/jessesquires/JSQMessagesViewController/blob/develop/CONTRIBUTING.md) and update your issue!

**Comment by armanshan12** [comment]
@jessesquires I have updated my issue. Thanks!

**Comment by jessesquires** [comment]
FYI: I know the firebase demo is a bit outdated by now

**Comment by armanshan12** [comment]
I also get this as I scroll, if it helps:

2015-10-04 17:17:10.047 HelloWorld[870:334106] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. Try this: (1) look at each constraint and try to figure out which you don't expect; (2) find the code that added the unwanted constraint or constraints and fix it. (Note: If you're seeing NSAutoresizingMaskLayoutConstraints that you don't understand, refer to the documentation for the UIView property translatesAutoresizingMaskIntoConstraints) 
(
    "<NSAutoresizingMaskLayoutConstraint:0x13690e4f0 h=--& v=--& V:[UIView:0x136cbb4c0(58)]>",
    "<NSLayoutConstraint:0x136cbb960 V:[JSQMessagesLabel:0x136cbb620(20)]>",
    "<NSLayoutConstraint:0x136cbc210 V:[JSQMessagesLabel:0x136cbbed0(20)]>",
    "<NSLayoutConstraint:0x136aca570 V:[JSQMessagesLabel:0x136aca280(20)]>",
    "<NSLayoutConstraint:0x136acac30 V:|-(0)-[JSQMessagesLabel:0x136cbb620]   (Names: '|':UIView:0x136cbb4c0 )>",
    "<NSLayoutConstraint:0x136acad20 V:[JSQMessagesLabel:0x136cbb620]-(0)-[JSQMessagesLabel:0x136cbbed0]>",
    "<NSLayoutConstraint:0x136acadc0 V:[JSQMessagesLabel:0x136cbbed0]-(0)-[UIView:0x136cbc590]>",
    "<NSLayoutConstraint:0x136acaf50 V:[JSQMessagesLabel:0x136aca280]-(0)-|   (Names: '|':UIView:0x136cbb4c0 )>",
    "<NSLayoutConstraint:0x136acafa0 V:[UIView:0x136cbc590]-(0)-[JSQMessagesLabel:0x136aca280]>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x136aca570 V:[JSQMessagesLabel:0x136aca280(20)]>

Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKit/UIView.h> may also be helpful.

**Comment by armanshan12** [comment]
I'm not sure why markdown deletes everything between the quotations.

**Comment by jessesquires** [comment]
hey @armanshan12 - this looks like a dupe of #817 

i suspect you are seeing crashes because you have a breakpoint turned on for exceptions.

if this is not the case, let me know and i'll re-open.

**Comment by armanshan12** [comment]
I do not have any breakpoints set. In fact, I have seen this crash on a released version of the app.

**Comment by armanshan12** [comment]
This is my observation: I start scrolling up and down, and tons of constraint warnings are printed out. I also observe lag on the device. Sometimes there is a big chunk of white space for at least a second and then everything is reloaded and looks normal, until i scroll again. Eventually the app crashes and prints out the error that is posted on first message of this thread.

**Comment by armanshan12** [comment]
@jessesquires 

I think I found the problem. The problem is with the avatars because they are being downloaded and set at a later time. If I remove the following code:

        SDWebImageManager.sharedManager().downloadImageWithURL(url, options: .LowPriority, progress: nil) { (image, error, _, _, _) -> Void in
            avatar.avatarImage = image
            avatar.avatarHighlightedImage = image
            self.collectionView!.reloadItemsAtIndexPaths([indexPath])
        }

in my avatarImageDataForItemAtIndexPath method, the placeholder avatars are used, and the page loads up without any lags or crashes. Constraint warning still print to console, but that is it. As soon as I add "self.collectionView!.reloadItemsAtIndexPaths([indexPath])" back to the code, the lag begins. What is the protocol for avatar images that are being downloaded asynchronously? Because if I don't add that line, the avatars are not switched from placeholder image to the actual image, unless I scroll up and down, in which case the cells are regenerated, and this time the avatar photos are in the cash, so they are set synchronously.

Thanks!

**Comment by jessesquires** [comment]
Hey @armanshan12  -- my comment at https://github.com/jessesquires/JSQMessagesViewController/issues/1217#issuecomment-145276545 basically describes the same issue/situation that you have. The `avatarImageDataForItemAtIndexPath:` is exactly like `messageDataForItemAtIndexPath:`.

Simply put, these `dataSource` and `delegate` methods are the wrong place for this. 

I would probably put the `SDWebImageManager` stuff in `viewDidLoad`. You should know what avatars you need to download there, and there are likely just 2 (or maybe a few more).

I also noticed you are creating a new `JSQMessagesAvatarImage` **every time** `avatarImageDataForItemAtIndexPath:` is called. This is a terrible idea. This isn't necessarily cheap, especially how you're doing it. I recommend following the pattern in the demo project. 

Here's how the flow should be:

1. Create and cache all `JSQMessagesAvatarImage` objects. Likely, this will just include a placeholder image.
* Start `SDWebImageManager` downloads in `viewDidLoad`
* On completion, update the appropriate `JSQMessagesAvatarImage` object by setting `avatarImage`
* Call `reloadData`, or only reload the visible cells, or grab `[collectionView visibleCells]` and update the avatars.

The library will *"do the right thing"* when it queries your data. If you have `avatarImage`, it will display. Otherwise, the placeholder is a fallback.

**Comment by armanshan12** [comment]
Hi @jessesquires,

The reason why I did not download the images at viewDidLoad is because this is a chat with possibly 100+ people, so it would not make sense to download every single person's image, if the user ends up not scrolling over all those people. I ended up moving the download method to cellForItemAtIndexPath, where I have access to the UIImageView, and it works correctly.

Thank you for all the help!
